### PR TITLE
random-search

### PR DESCRIPTION
**Pull request by pknut**

### CODE REVIEW

File: hyperparameter_tuning.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
+import gc
+
+import numpy as np
+from deepsense import neptune
```
Comment (kamil-kaczmarek) [inline hyperparameter_tuning.py]: neptune sits only in the callbacks

File: hyperparameter_tuning.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+from sklearn.externals import joblib
+
+from steps.base import BaseTransformer
+from utils import set_seed
+
+
+class RandomSearchOptimizer(BaseTransformer):
```
Comment (kamil-kaczmarek) [inline hyperparameter_tuning.py]: add docstring to this class. One have no idea what are these params.
Use Type hints.

File: hyperparameter_tuning.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    def __init__(self, TransformerClass, params,
+                 score_func, maximize,
+                 train_input_keys, valid_input_keys,
+                 n_runs,
+                 callbacks=[]):
```
Comment (kamil-kaczmarek) [inline hyperparameter_tuning.py]: do not use mutable type in the init. Replace `[]` with `None`

File: hyperparameter_tuning.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+        if self.maximize:
+            best_score, best_param_set = results_sorted[-1]
+        else:
+            best_score, best_param_set = results_sorted[0]
+
+        for callback in self.callbacks:
+            callback.on_search_end(results=results)
+
+        self.best_transformer = self.TransformerClass(**best_param_set)
```
Comment (kamil-kaczmarek) [inline hyperparameter_tuning.py]: Why do I have to inject Transformer inside, transformer? It seems that we are breaking our abstraction, here. What do you think @pknut ?

File: hyperparameter_tuning.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            run_score = self.score_func(y_valid, y_pred_valid_value)
+            results.append((run_score, param_set))
+
+            del y_pred_valid, transformer
+            gc.collect()
+
+            for callback in self.callbacks:
+                callback.on_run_end(score=run_score, params=param_set)
+
+        assert len(results) > 0, 'All random search runs failed, check your parameter space'
```
Comment (kamil-kaczmarek) [inline hyperparameter_tuning.py]: assertion error message should be more descriptive. Try to add some hint what to do

File: hyperparameter_tuning.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    def save(self, filepath):
+        self.best_transformer.save(filepath)
+
+    def load(self, filepath):
+        self.best_transformer.load(filepath)
+        return self
+
+
+def create_param_space(params, n_runs):
+    seed = np.random.randint(1000)
```
Comment (kamil-kaczmarek) [inline hyperparameter_tuning.py]: random seed should be global, not local

File: hyperparameter_tuning.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        elif mode == 'uniform':
+            value = np.random.uniform(low=range_min, high=range_max)
+        elif mode == 'log-uniform':
+            value = np.exp(np.random.uniform(low=np.log(range_min), high=np.log(range_max)))
+        else:
+            raise NotImplementedError
+    return value
+
+
+class GridSearchCallback:
```
Comment (kamil-kaczmarek) [inline hyperparameter_tuning.py]: All callbacks should be placed in single Python file. We should discuss where to place them (in the steppy-toolkit).

File: hyperparameter_tuning.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    def on_run_end(self, score, params):
+        return NotImplementedError
+
+    def on_search_end(self, results):
+        return NotImplementedError
+
+
+class NeptuneMonitor(GridSearchCallback):
```
Comment (kamil-kaczmarek) [inline hyperparameter_tuning.py]: Similarly to the previous callback, let's discuss where to place them.

File: hyperparameter_tuning.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        self.ctx.channel_send('lowest score', x=0, y=lowest_score)
+
+        self.highest_params_channel.send(y=highest_param_set)
+        self.lowest_params_channel.send(y=lowest_param_set)
+
+    def _create_text_channel(self, name=''):
+        return self.ctx.create_channel(name=name, channel_type=neptune.ChannelType.TEXT)
+
+
+class SaveResults(GridSearchCallback):
```
Comment (kamil-kaczmarek) [inline hyperparameter_tuning.py]: What is this class for? Is it Transformer?

File: hyperparameter_tuning.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+from utils import set_seed
+
+
+class RandomSearchOptimizer(BaseTransformer):
+    def __init__(self, TransformerClass, params,
+                 score_func, maximize,
+                 train_input_keys, valid_input_keys,
+                 n_runs,
+                 callbacks=[]):
+        self.TransformerClass = TransformerClass
```
Comment (kamil-kaczmarek) [inline hyperparameter_tuning.py]: Here, it seems that you inject transformer inside the transformer. I think we should refactor this in a way that it does not requires this trick

File: hyperparameter_tuning.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+    def fit(self, **kwargs):
+        if self.train_input_keys:
+            train_inputs = {input_key: kwargs[input_key] for input_key in self.train_input_keys}
+        else:
+            train_inputs = kwargs
+        X_valid, y_valid = kwargs[self.valid_input_keys[0]], kwargs[self.valid_input_keys[1]]
+
+        results = []
+        for i, param_set in enumerate(self.param_space):
```
Comment (kamil-kaczmarek) [inline hyperparameter_tuning.py]: `i` is not used

File: neptune_random_search.yaml [hunk 12]
```diff
@@ -0,0 +1,55 @@
+project: HC
+
+name: home-credit
+tags: [solution]
```
Comment (kamil-kaczmarek) [inline neptune_random_search.yaml]: change to solution-1

File: neptune_random_search.yaml [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+# Validation
+  target_bins: 100
+  validation_size: 0.2
+
+# Execution
+  overwrite: 1
+  num_workers: 10
+  verbose: 1
+
+  # Preprocessing
```
Comment (kamil-kaczmarek) [inline neptune_random_search.yaml]: no spaces before `# Preprocessing`

File: neptune_random_search.yaml [hunk 14]
```diff
@@ -0,0 +1,55 @@
+project: HC
+
+name: home-credit
```
Comment (kamil-kaczmarek) [inline neptune_random_search.yaml]: change to `home-credit-default-risk`
